This project aims to clean the a sample of the data obtained from a Kaggle dataset.  This dataset was originally scraped from eBay Kleinanzeigen, a section of the eBay website.

The sample in question has been "dirtied" as part of a guided project for dataquest.io in which I will re-clean the dataset and perform some analysis. 

In [2]:
import numpy as np
import pandas as pd
autos = pd.read_csv("autos.csv", encoding="Latin-1")
#Utf-8 generated errors on reading

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Generally all text data is primarily in german and the majority of columns appear to have been read as objects. Due to dataset.info() showing "non-null, we can deduce that although the data does contain NaN or null vaules this is not more than 20% per column.

Most dates appear to have been read into the datetime format. This exception to this is the year and month of regsitration which are in two separate columns as integer values.

"price" and "odometer" data contains units, so this will have to be extracted in included in the column heading so that numerical operations can be carried out.

The name column will be hard to extract any data from as it appears to have been scraped from a free text field.

The column names are camel rather than snake case, so we will need to correct this first to aid in working on the data.

In [5]:
corrected_cols=['date_crawled', 'name', 'seller', 'offer_type', 'price',\
                'abtest','vehicle_type', 'registration_year', 'gearbox', \
                'power_ps', 'model','odometer', 'registration_month', \
                'fuel_Type', 'brand','unrepaired_damage', 'ad_created', \
                'num_pictures', 'post_code','last_seen']

As seen above I have made the following changes to the column headings.
Generally where possible I have presevered the existing name and simply converted to snake_case. I have included more detail on specific changes below.
yearOfRegistration to registration_year: this simplifies and shortens the heading
monthOfRegistration to registration_month: change for same reason as above
notRepairedDamage to unrepaired_damage : improved syntax
dateCreated to ad_created : clarifying the difference between the datum scraped and the ad it was scrapted from

In [6]:
autos.columns = corrected_cols
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_Type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_pictures  post_code  \
0     peugeot              nein  2016-03-26 00:00:00             0      79588   
1         bmw              nein  2016-04-04 00:00:00             0      71034   
2  volkswagen              nein  2016-03-26 00:00:00             0      35394   
3       smart              nein  2016-03-12 00:00:00             0      33729   
4        ford              nein  2016-04-01 00:00:00             0      39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

As mentioned above the price and odometer columns require some work as they are currently stored as text.  
The price column contains a dollar sign and commas separating the thousands.  I shall remove these and then change the data type to in64.
The odometer data contains km units and similar commas separating thousands.

After these transformations I will rename the columns to contain units.

In [7]:
autos['price'] = (autos['price']
                  .str.replace("$","")
                  .str.replace(",","")
                  .astype("int64"))

autos['odometer'] = (autos['odometer']
                  .str.replace("km","")
                  .str.replace(",","")
                  .astype("int64"))

In [8]:
autos.rename(columns={'odometer':'odometer_km','price':'price_usd'}, inplace=True)


I will also need to check the date created column to ascertain whether this has been imported as a datetime field correctly.

In [9]:
#autos['ad_created'] = pd.to_datetime(autos['ad_created'])
#autos['ad_created'] = pd.to_datetime(autos['date_crawled'])
#autos['ad_created'] = pd.to_datetime(autos['last_seen'])


Upon investigation I discovered that the ad_created column had not imported as a datetime.  As such I converted this so that any trends can be investigated in these data

In [10]:
autos['seller'].value_counts()
autos['offer_type'].value_counts()
autos['unrepaired_damage'].value_counts()
autos['num_pictures'].value_counts()

0    50000
Name: num_pictures, dtype: int64

I decided to look into the following columns as in the summary data they appear to mostly or entirely contain 1 value: seller, offer_type, unrepaired_damage and num_pictures.

After looking at the data present in these columns I believe it would be pertinent to drop the following columns: seller, offer_type and num_pictures

In [11]:
autos.drop(['seller', 'offer_type', 'num_pictures'],axis=1,inplace=True)
autos.describe()

price_usd  registration_year      power_ps    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month     post_code  
count        50000.000000  50000.000000  
mean             5.723360  50813.627300  
std              3.711984  25779.747957  
min              0.000000   1067.000000  
25%              3.000000  30451.000000  
50%              6.000000  49577.000000  
75%              9.000000  71540.000000  
max             12.000000  99998.000000

Looking at the above .describe() output the odometer data appears consistent however the max value for price is \$10,000,000 where the mean is \$9840.  I believe this is anomolous data and will need to remove this before further analysis  can be carried out. 

In [12]:
price = autos['price_usd']
price.value_counts().sort_index(ascending=False).head(13)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
Name: price_usd, dtype: int64

In [13]:
autos[autos['price_usd'].between(500000,99999999)].sort_values(by='price_usd',ascending=False)

date_crawled                                               name  \
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   

       price_usd   abtest vehicle_type  registration_year    gearbox  \
39705   99999999  control    limousine               1999  automatik   
42221   27322222  control    limousine               2014    manuell   
27371   12345678  control          NaN               2017        NaN   
39377   12345678  control          NaN               2018    manuell   
47598   12345678  control    limousine               2001    manuell   
2897    11111111     test    limousine               1973    manuell   
24384   11111111     test          NaN               1995        NaN   
11137   10000000  control        coupe               1960    manuell   
47634    3890000     test        coupe               2006        NaN   
7814     1300000  control        coupe               1992        NaN   
22947    1234566  control        kombi               1999  automatik   
514       999999     test        kombi               2009    manuell   
43049     999999     test          bus               1981    manuell   
37585     999990     test    limousine               1985    manuell   

       power_ps        model  odometer_km  registration_month fuel_Type  \
39705       224     s_klasse       150000                   9    benzin   
42221       163           c4        40000                   2    diesel   
27371        95        punto       150000                   0       NaN   
39377        95          v40       150000                   6       NaN   
47598       101       vectra       150000                   3    benzin   
2897         48       escort        50000                   3    benzin   
24384         0          NaN       150000                   0       NaN   
11137       368          NaN       100000                   1    benzin   
47634       799          NaN         5000                   7       NaN   
7814          0          NaN        50000                  12       NaN   
22947       190          NaN       150000                   2    diesel   
514         101        focus       125000                   4    benzin   
43049        70  transporter       150000                   1    benzin   
37585       111        jetta       150000                  12    benzin   

                brand unrepaired_damage           ad_created  post_code  \
39705   mercedes_benz               NaN  2016-03-22 00:00:00      73525   
42221         citroen               NaN  2016-03-08 00:00:00      76532   
27371            fiat               NaN  2016-03-09 00:00:00      96110   
39377           volvo              nein  2016-03-08 00:00:00      14542   
47598            opel              nein  2016-03-31 00:00:00       4356   
2897             ford              nein  2016-03-12 00:00:00      94469   
24384      volkswagen               NaN  2016-03-21 00:00:00      

Looking at the data in the price_usd column for sales greater than $500,000 there appears to be a mixture of erroneous and real data.  

Rows 39705 and 42221 can be dropped immediately as they are not sales of cars.  Their titles translate to exchanges or lease takeovers. 

The following rows appear to be sales but the dollar value does not correspond to the expected value for these vehicles: 27371 39377 47598 2897 24384 22947 514 43049 37585 I will drop these rows for now also.

11137 47634 and 7814 could be legitimate sales as these relate to supercars.

In [14]:
drops = [39705,42221,27371,39377,47598,2897,24384,22947,514,43049,37585]

autos.drop(drops,inplace=True)


In [15]:
autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025386
2016-03-06    0.013943
2016-03-07    0.035968
2016-03-08    0.033267
2016-03-09    0.033207
2016-03-10    0.032127
2016-03-11    0.032487
2016-03-12    0.036768
2016-03-13    0.015563
2016-03-14    0.036628
2016-03-15    0.033987
2016-03-16    0.029506
2016-03-17    0.031507
2016-03-18    0.013063
2016-03-19    0.034908
2016-03-20    0.037828
2016-03-21    0.037488
2016-03-22    0.032907
2016-03-23    0.032387
2016-03-24    0.029106
2016-03-25    0.031747
2016-03-26    0.032487
2016-03-27    0.031047
2016-03-28    0.034848
2016-03-29    0.034168
2016-03-30    0.033627
2016-03-31    0.031907
2016-04-01    0.033807
2016-04-02    0.035408
2016-04-03    0.038689
2016-04-04    0.036528
2016-04-05    0.013103
2016-04-06    0.003181
2016-04-07    0.001420
Name: date_crawled, dtype: float64

The date_crawled column appears to be evenly distributed.  This would appear to indicate that there is no bias towards day of the month/day of the week when the scraper picked up the ad listing.

In [16]:
(autos['ad_created']
 .str[:10].value_counts(normalize=True,dropna=False)
 .sort_index()
 .tail(34)
 .sum())

0.9958790934005481

The ad_created series contains a wider range of dates represented due to the way that the sampling was carried out.  However looking at just the 34 day period overwhich the data was scraped,  it would appear that ~99% of ads were created in this time, and as such the ads outside of this period can potentially be dropped.

In [17]:
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index()
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index().tail(5).sum()

0.5262357718698113

This field indicates one of two things: that the listing came to an end and thus wasn't picked up by the scaper again, or that the listing ended outside of the sampling period.  Of these two it would appear that the latter is the more significant as ~50% of the listings were "last seen" by the scaper within the last 5 days of the sampling period.

In [18]:
autos['registration_year'].describe()

count    49989.000000
mean      2005.074576
std        105.724197
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Looking at the registration_year data there would appear to be some anomolous figures given that the max is the year 9999 and the min is 1000.  The data set also appears to be right skewed. 

As the first cars were registered in the early decades of the 20th century it would be prudent to assume that any listings with the car registered prior to this date can be dropped.  Similarly this survey was carried out in 2016, and so any years of registration after this can also safely be dropped.

In [19]:
year_drops = autos.loc[~autos["registration_year"].between(1900,2016),:]
autos = autos.drop(year_drops.index)
autos["registration_year"].value_counts(normalize=True).sort_index()

1910    0.000187
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000042
1937    0.000083
1938    0.000021
1939    0.000021
1941    0.000042
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000042
1952    0.000021
1953    0.000021
1954    0.000042
1955    0.000042
1956    0.000104
1957    0.000042
1958    0.000083
1959    0.000146
1960    0.000708
1961    0.000125
1962    0.000083
1963    0.000187
1964    0.000250
1965    0.000354
1966    0.000458
1967    0.000562
1968    0.000541
          ...   
1987    0.001562
1988    0.002957
1989    0.003769
1990    0.008226
1991    0.007414
1992    0.008143
1993    0.009267
1994    0.013745
1995    0.027323
1996    0.030071
1997    0.042233
1998    0.051084
1999    0.062434
2000    0.069847
2001    0.056269
2002    0.052750
2003    0.056790
2004    0.056998
2005    0.062788
2006    0.056394
2007    0.047981
2008    0.046461
2009    0.043670
2010    0.033258
2011    0.034028
2012    0.027552
2013    0.016785
2014    0.0138

After removing the outliers the data now appears as expected.  The listings are heavily biased towards recent registration years.  Roughly the last 20 years back to 1996.

I will now investigate the various brands in this dataset to look into which brand of car sold for the most on eBay over this sample period.

I will do this by aggregating the data for the brands of car with greater than 5% of the listings in this data.  This gives the following brands: VW, BMW, Opel, Mercedes, Audi, Ford, Renault.  These are the most common car brands are so are the most relevant to look into the prices of.

In [64]:
mask_5per = autos["brand"].value_counts(normalize=True)>=.04
values_agg = mask_5per[mask_5per==True].index
price_per_brand = {}

for brand in values_agg:
    mean_price = np.floor(autos.loc[autos["brand"]==brand,"price_usd"].mean())
    price_per_brand[brand]=mean_price
    
price_per_brand



{'audi': 9093.0,
 'bmw': 8102.0,
 'ford': 3652.0,
 'mercedes_benz': 8485.0,
 'opel': 2876.0,
 'renault': 2395.0,
 'volkswagen': 5231.0}

As shown above the mean prices for the selected brands of cars are ranked as follows.

Audi
Mercedes Benz
BMW
VW
Ford
Opel
Renault

As these are used car sales this could indicate that these brands of cars hold their value better.  However it could also indicate that these cars are also "sold newer" than other brands and so their condition at sale is better.

In order to ascertain which is true I will need to look into the registration dates or mileage for these brands also.  Higher average registration years for the brands would confirm that these are _sold newer_, likewise for mileage